Add test for models

ankitkataria opened the pull request:
Added rails-rspec, factory-bot and faker dep
Tests for model -
- [x] group
- [x] assignment
- [x] collaboration
- [x] group_member
- [x] pending_invitation
- [x] project
- [x] star
- [x] user

Todo - 
- [x] add rake task for test
- [x] update readme with test instructions

Fixes #134

ankitkataria commented:
@vik-y How does the PR look so far? Should I continue adding specs for all the models like I've done for group and assignments currently?

ankitkataria commented:
No I've only been using pg for dev and test. I've been intending to ask why
there is an option for MySQL.

On Thu 28 Feb, 2019, 1:41 PM Aboobacker MK, <notifications@github.com>
wrote:

> *@tachyons* commented on this pull request.
> ------------------------------
>
> In Gemfile
> <https://github.com/CircuitVerse/CircuitVerse/pull/148#discussion_r261069773>
> :
>
> > @@ -70,6 +70,9 @@ group :development, :test do
>    gem 'capybara', '~> 2.13'
>    gem 'selenium-webdriver'
>    gem 'pry-rails'
> +  gem 'pg', '~> 1.0.0'
>
> Using mysql and Pg together ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/CircuitVerse/pull/148#pullrequestreview-208918862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAEHiF-MICTejaaTzcnko-1XXrEe02rks5vR48UgaJpZM4bWHF0>
> .
>

vik-y commented:
@ankitkataria Looks perfect so far. Great job and thanks a ton for putting it together. 

Please go ahead and do the changes you wish to, I'll look into it in detail by Saturday and then let's try to merge this.

ankitkataria commented on Gemfile:
```diff
@@ -70,6 +70,9 @@ group :development, :test do
   gem 'capybara', '~> 2.13'
   gem 'selenium-webdriver'
   gem 'pry-rails'
+  gem 'pg', '~> 1.0.0'
```
@tachyons Is there a specific reason for using mysql in development and pg in production? I've been using only pg for both dev and testing. Should I be changing this to use only mysql?

tachyons replied:
No, but it seems that we have both pg and mysql requirement in the Gemfile. It is better to get rid of one cc @satu0king

ankitkataria replied:
If I remove `pg` from here, I'll have to move `gem 'mysql2' , '>= 0.3.18' , '<0.5'` under `group :development, :test` right?

ankitkataria commented:
@vik-y Please review! :gift:

tachyons commented:
Good initiative. 
1. It seems that you are testing mostly rails behaviours like callbacks and associations. Those behaviors are already tested in rails project. (You can keep shoulda specs)

2. Model methods seems stubbed instead of testing them. Write tests for all important public methods in models 

3. When testing validations, make sure that you are not getting false positives. Eg

```ruby
it "user invalid with invalid email" do
  expect(create (:user)).to be_valid
  expect(create (:user, email: 'inVaLid')).not_to be_valid
end
```

Here first expectation is to make sure that factory was valid by default

3. Use coverall for code coverage

tachyons commented on spec/factories/user.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :user do
+    email { "test@testmail.com" }
```
Use faker

tachyons commented on spec/factories/user.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :user do
+    email { "test@testmail.com" }
+    password { "password" }
```
Use faker

tachyons commented on spec/models/assignment_spec.rb:
```diff
@@ -0,0 +1,27 @@
+      allow_any_instance_of(Assignment).to receive(:set_deadline_job)
+    end
+
+    it "should all respective callbacks" do
```
You mean "should fire all respective callbacks" ? Anyway this tests seems unnecessary since this is the default rails behavior

tachyons commented on spec/models/project_spec.rb:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    it "doesn't validate with public access type" do
+        project = FactoryBot.build(:project, assignment: @assignment, author: @user, project_access_type: "Public")
```
2 space indentation ?

ankitkataria commented:
@tachyons Thanks for the review. :smile: 

> Use coverall for code coverage

Will implement this. 

> When testing validations, make sure that you are not getting false positives. Eg

Will get this fixed. 

> Model methods seems stubbed instead of testing them. Write tests for all important public methods in models

In this PR I was aiming to provide the foundation for the test suite which could be used as the basis for developing it further. I hoping to get this PR merged first and once the foundation was set, was aiming to improve this test suite as a separate smaller PR. This way reviewing could also be easier. 
Will this be a good idea?

tachyons commented:
Sure 👍

ankitkataria commented:
Thanks :smile: !
I'll implement all the other changes by the end of the day.

ankitkataria commented:
@tachyons I've worked on points in your review. I've added coveralls for test coverage. For now, I've removed the functions stubs but have kept the tests for callbacks. I'll add the test for the public methods of the models following this PR.

ankitkataria commented:
@vik-y Is this good to go? :sweat_smile:

satu0king commented:
Thank you for this much-needed PR. Also @tachyons, thank you for the reviews! I will merge this after @vik-y approves.

armudgal commented:
Can we squash the commits here please, If I am not wrong we can squash the commits into two(maybe three) commits (one for test, one for readme and code coverage) but feel free to reduce it to the number you feel should be apt.

ankitkataria commented:
@armudgal I've divided the addition of tests to 2 commits, and have reduced the number of commits. Thanks for the suggestion.
I hope this can be merged quickly, so that I can start working on Issue #167. :slightly_smiling_face:

armudgal commented:
I can still see 7 commits, is that intended and is the reduced number of commits?

ankitkataria commented:
Earlier there were 12. 
I've divided them into adding deps, adding factories, adding specs(2 commits for these), adding rake task, coveralls and readme.

armudgal commented:
Okay, this looks LGTM, @vik-y can you take a look at this so we can go ahead merging. Thanks.

tachyons commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,106 @@
+require "coveralls"
+require "paperclip/matchers"
+
+Coveralls.wear!
```
```suggestion
Coveralls.wear!('rails')
```

vik-y commented on Gemfile:
```diff
@@ -64,16 +64,21 @@ gem 'will_paginate-bootstrap'
 # gem 'capistrano-rails', group: :development
 
 group :development, :test do
+  gem 'mysql2' , '>= 0.3.18' , '<0.5'
```
@ankitkataria We are using postgres in production,I would prefer it for tests as well.

ankitkataria replied:
Locally, I was using postgres for it too, use pg is as easy as `gem 'pg'` in the `:test` group and making necessary changes in  `database.yml`. I'll add an option for the same in the upcoming prs.

vik-y commented on README.md:
````diff
@@ -75,13 +80,13 @@ production:
 - If you are facing difficulties installing RVM, most probably it is because of an older version of rvm shipped with Ubuntu's desktop edition and updating the same resolves the problem.
 
   Removing RVM
-  ```  
````
@ankitkataria Which IDE are you using? Why does add the "blank" at the end of multiple lines?

ankitkataria replied:
Those blank spaces were already present in the README. I use sublime-text 3 along with a editorconfig that removed the blank spaces in the end.

ankitkataria commented:
@vik-y Thanks for the review! I'll take care all the mentioned points in the upcoming improvement pr.
